libz-rs-sys/src/gz.rs

```diff
@@ -0,0 +1,239 @@
+// public gzFile_s part at the front for ABI compatibility).
+#[repr(C)]
+struct GzState {
+    x: gzFile_s,         // "x" for the exposed part of the data structure
+
+    // Fields used for both reading and writing
+    mode: GzMode,
+    fd: c_int,           // file descriptor
+    path: *const c_char,
+    size: usize,         // buffer size; can be 0 if not yet allocated
+    want: usize,         // requested buffer size
+    input: *mut Bytef,   // input buffer
+    output: *mut Bytef,  // output buffer
+    direct: bool,        // true in pass-through mode, false if processing gzip data
+
+    // Fields used just for reading
+    // FIXME: add the 'how' field when read support is implemented
+    start: i64,
+    eof: bool,           // whether we have reached the end of the input file
+    past: bool,          // whether a read past the end has been requested
+
+    // Fields used just for writing
+    level: i8,
+    strategy: Strategy,
+    reset: bool,         // whether a reset is pending after a Z_FINISH
+
+    // Fields used for seek requests
+    skip: i64,           // amount to skip (already rewound if backwards)
+    seek: bool,          // whether a seek request is pending
+
+    // Error information
+    err: c_int,          // last error (0 if no error)
+    msg: *const c_char,  // error message from last error (NULL if none)
+
+    // FIXME: add the zstream field when read/write support is implemented
+}
+
+// Gzip operating modes
+// NOTE: These values match what zlib-ng uses.
+#[derive(Eq, PartialEq)]
+enum GzMode {
+    GZ_NONE = 0,
+    GZ_READ = 7247,
+    GZ_WRITE = 31153,
+    GZ_APPEND = 1,
+}
+
+const GZBUFSIZE: usize = 128 * 1024;
+
+#[cfg(feature = "rust-allocator")]
+const ALLOCATOR: &Allocator = &Allocator::RUST;
+
+#[cfg(not(feature = "rust-allocator"))]
+#[cfg(feature = "c-allocator")]
+const ALLOCATOR: &Allocator = &Allocator::C;
+
+#[cfg(not(feature = "rust-allocator"))]
+#[cfg(not(feature = "c-allocator"))]
+compile_error!("Either rust-allocator or c-allocator feature is required");
+
+/// Open a gzip file for reading or writing.
+///
+/// # Safety
+/// The caller must ensure that path and mode point to valid C strings. If the
+/// return value is non-NULL, caller must delete it using only [`gzclose`].
+#[cfg_attr(feature = "export-symbols", export_name = crate::prefix!(gzopen))]
+pub unsafe extern "C-unwind" fn gzopen(path: *const c_char, mode: *const c_char) -> gzFile {
+    gzopen_help(path, -1, mode)
+}
+
+/// Internal implementation shared by gzopen and gzdopen.
+///
+/// # Safety
+/// The caller must ensure that path and mode are NULL or point to valid C strings.
+unsafe fn gzopen_help(path: *const c_char, fd: c_int, mode: *const c_char) -> gzFile {
+    if path.is_null() || mode.is_null() {
+        return ptr::null_mut();
+    }
+
+    let Some(state) = ALLOCATOR.allocate_zeroed(size_of::<GzState>()) else {
+        return ptr::null_mut();
+    };
+    let state = state.cast::<GzState>().as_mut();
```

Comment:
I think this is technically unsound because alignment is ignored. (it works out though because malloc aligns to at least 8 in practice). I'll refactor the zeroed allocator apis a bit to make this easier.

Reply:
https://github.com/trifectatechfoundation/zlib-rs/pull/327 should make this `allocator.allocate_zeroed_raw::<GzState>()` and will correctly handle alignment

Reply:
Thank you! And to allocate space for a C string, I guess it's `allocator.allocate_slice_raw::<u8>(strlen + 1)`?

Reply:
yes. Then you just need to make sure to write to the memory before you read it, and must write the `NULL` byte. We could add a variant for zeroed slices but I think `slice_allocate_raw` should work fine.